core/src/BitSource.cpp

```diff
@@ -24,54 +24,68 @@
 
 namespace ZXing {
 
-int
-BitSource::available() const
+static int ReadBitsImpl(int numBits, const ByteArray& bytes, int& byteOffset, int& bitOffset)
```

Comment:
ReadBitsImpl is a new function, not a member function belonging to the BitSource class.  I've checked the rest of the project and leading underscores are not used for function parameters so I think what I've done is correct.

Reply:
The reason `peakBits` is required is that the terminator code is longer than the mode codes. In QR codes the mode codes and the terminator code are all 4 bits.  This is great because we can read 4 bits from the data stream to see if the mode has changed or we should terminate.  In micro QR codes the mode codes vary in length but the terminator code is always longer than the mode code.
- M1 - 0 length mode code, 3 bits terminator code
- M2 - 1 bit mode code, 5 bits terminator code
- M3 - 2 bit mode code, 7 bits terminator code
- M4 - 3 bit mode code, 9 bits terminator code

The numeric mode code is 1, 2 or 3 0s and the terminator code is 3, 5, 7, or 9 0s.  When reading the bit stream we need to peak ahead to see if we're about to parse a terminator code or just a numeric code.  If the peak is not a terminator code then it will read the bits as normal.

Reply:
Thanks. Very informative. Could you maybe put this info into a comment at an appropriate place, now that it is readily available?